exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/Metadata.java

```diff
@@ -495,31 +499,65 @@ private ParquetFileMetadata_v3 getParquetFileMetadata_v3(ParquetTableMetadata_v3
    * @param p
    * @throws IOException
    */
-  private void writeFile(ParquetTableMetadata_v3 parquetTableMetadata, Path p) throws IOException {
+  private void writeFile(ParquetTableMetadata_v3 parquetTableMetadata, String path) throws IOException {
     JsonFactory jsonFactory = new JsonFactory();
     jsonFactory.configure(Feature.AUTO_CLOSE_TARGET, false);
     jsonFactory.configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false);
     ObjectMapper mapper = new ObjectMapper(jsonFactory);
     SimpleModule module = new SimpleModule();
     module.addSerializer(ColumnMetadata_v3.class, new ColumnMetadata_v3.Serializer());
     mapper.registerModule(module);
-    FSDataOutputStream os = fs.create(p);
+
+    // If multiple clients are updating metadata cache file concurrently, the cache file
+    // can get corrupted. To prevent this, write to a unique temporary file and then do
+    // atomic rename.
+    UUID randomUUID =  UUID.randomUUID();
+    Path tmpPath = new Path(path, new String(METADATA_FILENAME + "." + randomUUID));
+
+    FSDataOutputStream os = fs.create(tmpPath);
     mapper.writerWithDefaultPrettyPrinter().writeValue(os, parquetTableMetadata);
     os.flush();
     os.close();
+
+    // Use fileContext API as FileSystem rename is deprecated.
+    FileContext fileContext = FileContext.getFileContext(tmpPath.toUri());
+    Path finalPath = new Path(path, METADATA_FILENAME);
+
+    try {
+      fileContext.rename(tmpPath, finalPath, Options.Rename.OVERWRITE);
+    } catch (Exception e) {
+      logger.info("Rename from {} to {} failed", tmpPath.toString(), finalPath.toString(), e);
+    }
```

Comment:
Should we add a `finally` section which removes the temp file if it exists i.e. case rename failed?